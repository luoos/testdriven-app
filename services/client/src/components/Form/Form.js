import React from 'react';

const Form = (props) => {
  return (
    <div>
      {props.formType === 'Login' &&
        <h1 className="title is-1">Log In</h1>
      }
      {props.formType === 'Register' &&
        <h1 className="title is-1">Register</h1>
      }
      <hr/><br/>
      <form onSubmit={(event) => props.handleUserFormSubmit(event)}>
        {props.formType === 'Register' &&
          <div className="field">
            <label className="label is-large">Username</label>
            <div className="control has-icons-left">
              <input
                name="username"
                className="input is-large"
                type="text"
                placeholder="Enter a username"
                required
                value={props.formData.username}
                onChange={props.handleFormValueChange}
              />
              <span className="icon is-small is-left">
                <i className="fas fa-user"></i>
              </span>
            </div>
          </div>
        }
        <div className="field">
          <label className="label is-large">Email</label>
          <div className="control has-icons-left">
            <input
              name="email"
              className="input is-large"
              type="email"
              placeholder="Enter an email address"
              autoComplete="email"
              required
              value={props.formData.email}
              onChange={props.handleFormValueChange}
            />
            <span className="icon is-left">
              <i className="fas fa-envelope"></i>
            </span>
          </div>
        </div>
        <div className="field">
          <label className="label is-large">Password</label>
          <div className="control has-icons-left">
            <input
              name="password"
              className="input is-large"
              type="password"
              placeholder="Enter a password"
              autoComplete="current-password"
              required
              value={props.formData.password}
              onChange={props.handleFormValueChange}
            />
            <span className="icon is-left">
              <i className="fas fa-lock"></i>
            </span>
          </div>
        </div>
        <br/>
        <input
          type="submit"
          className="button is-primary is-large is-fullwidth"
          value="submit"
        />
      </form>
    </div>
  );
};

export default Form;