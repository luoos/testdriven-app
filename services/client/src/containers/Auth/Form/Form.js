import React from 'react';
import { connect } from 'react-redux';

import { updateObject } from '../../../shared/utility';
import * as actions from '../../../store/actions/index';

class Form extends React.Component {
  state = {
    formData: {
      username: '',
      email: '',
      password: ''
    }
  }

  componentDidMount() {
    this.clearForm();
  }

  clearForm = () => {
    this.setState({
      formData: {username: '', email: '', password: ''}
    });
  }

  handleFormValueChange = (event) => {
    const updatedForm = updateObject(this.state.formData,
      {[event.target.name]: event.target.value});
    this.setState({formData: updatedForm});
  }

  handleUserFormSubmit = (event) => {
    event.preventDefault();
    const formType = window.location.href.split('/').reverse()[0];
    let data = {...this.state.formData};
    if (formType === 'register') {
      this.props.onAuth(data.email, data.password, true, data.username);
    } else if (formType === 'login') {
      this.props.onAuth(data.email, data.password, false);
    } else {
      console.error(`No action for ${formType}`);
    }
    this.clearForm();
  }

  render() {
    return (
      <div>
        {this.props.formType === 'Login' &&
          <h1 className="title is-1">Log In</h1>
        }
        {this.props.formType === 'Register' &&
          <h1 className="title is-1">Register</h1>
        }
        <hr/><br/>
        <form onSubmit={(event) => this.handleUserFormSubmit(event)}>
          {this.props.formType === 'Register' &&
            <div className="field">
              <label className="label is-large">Username</label>
              <div className="control has-icons-left">
                <input
                  name="username"
                  className="input is-large"
                  type="text"
                  placeholder="Enter a username"
                  required
                  value={this.state.formData.username}
                  onChange={this.handleFormValueChange}
                />
                <span className="icon is-small is-left">
                  <i className="fas fa-user"></i>
                </span>
              </div>
            </div>
          }
          <div className="field">
            <label className="label is-large">Email</label>
            <div className="control has-icons-left">
              <input
                name="email"
                className="input is-large"
                type="email"
                placeholder="Enter an email address"
                autoComplete="email"
                required
                value={this.state.formData.email}
                onChange={this.handleFormValueChange}
              />
              <span className="icon is-left">
                <i className="fas fa-envelope"></i>
              </span>
            </div>
          </div>
          <div className="field">
            <label className="label is-large">Password</label>
            <div className="control has-icons-left">
              <input
                name="password"
                className="input is-large"
                type="password"
                placeholder="Enter a password"
                autoComplete="current-password"
                required
                value={this.state.formData.password}
                onChange={this.handleFormValueChange}
              />
              <span className="icon is-left">
                <i className="fas fa-lock"></i>
              </span>
            </div>
          </div>
          <br/>
          <input
            type="submit"
            className="button is-primary is-large is-fullwidth"
            value="submit"
          />
        </form>
      </div>
    );
  }
};

const mapDispatchToProps = dispatch => {
  return {
    onAuth: (email, password, isSignup, username) => dispatch(
      actions.auth(email, password, isSignup, username)
    ),
  }
}

export { Form }; // work around

export default connect(null, mapDispatchToProps)(Form);