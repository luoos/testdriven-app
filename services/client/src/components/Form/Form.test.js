import React from 'react';
import { shallow } from 'enzyme';
import renderer from 'react-test-renderer';

import Form from './Form';

const testData = [
  {
    formType: 'Register',
    title: 'Register',
    formData: {
      username: '',
      email: '',
      password: ''
    },
    handleUserFormSubmit: jest.fn(),
    handleFormValueChange: jest.fn(),
  },
  {
    formType: 'Login',
    title: 'Log In',
    formData: {
      email: '',
      password: ''
    },
    handleUserFormSubmit: jest.fn(),
    handleFormValueChange: jest.fn(),
  }
]

describe('<Form />', () => {

  testData.forEach((el) => {
    const component = <Form {...el} />;
    it(`${el.formType} Form renders properly`, () => {
      const wrapper = shallow(component);
      const h1 = wrapper.find('h1');
      expect(h1.length).toBe(1);
      expect(h1.get(0).props.children).toBe(el.title);
      const formGroup = wrapper.find('.control');
      expect(formGroup.length).toBe(Object.keys(el.formData).length);
      expect(formGroup.get(0).props.children[0].props.name).toBe(Object.keys(el.formData)[0]);
      expect(formGroup.get(0).props.children[0].props.value).toBe('');
    });

    it(`${el.formType} Form renders a snapshot properly`, () => {
      const tree = renderer.create(component).toJSON();
      expect(tree).toMatchSnapshot();
    })

    it(`${el.formType} Form submits the form properly`, () => {
      const wrapper = shallow(component);
      const input = wrapper.find('input[type="email"]');
      expect(el.handleUserFormSubmit).toHaveBeenCalledTimes(0);
      expect(el.handleFormValueChange).toHaveBeenCalledTimes(0);
      input.simulate('change');
      expect(el.handleFormValueChange).toHaveBeenCalledTimes(1);
      wrapper.find('form').simulate('submit', el.formData)
      expect(el.handleUserFormSubmit).toHaveBeenCalledWith(el.formData);
      expect(el.handleUserFormSubmit).toHaveBeenCalledTimes(1);
    })
  })

})
