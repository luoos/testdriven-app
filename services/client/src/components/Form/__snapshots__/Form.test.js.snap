// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Form /> Login Form renders a snapshot properly 1`] = `
<div>
  <h1
    className="title is-1"
  >
    Log In
  </h1>
  <hr />
  <br />
  <form
    onSubmit={[Function]}
  >
    <div
      className="field"
    >
      <label
        className="label is-large"
      >
        Email
      </label>
      <div
        className="control has-icons-left"
      >
        <input
          className="input is-large"
          name="email"
          placeholder="Enter an email address"
          required={true}
          type="email"
          value=""
        />
        <span
          className="icon is-left"
        >
          <i
            className="fas fa-envelope"
          />
        </span>
      </div>
    </div>
    <div
      className="field"
    >
      <label
        className="label is-large"
      >
        Password
      </label>
      <div
        className="control has-icons-left"
      >
        <input
          className="input is-large"
          name="password"
          placeholder="Enter a password"
          required={true}
          type="password"
          value=""
        />
        <span
          className="icon is-left"
        >
          <i
            className="fas fa-lock"
          />
        </span>
      </div>
    </div>
    <br />
    <input
      className="button is-primary is-large is-fullwidth"
      type="submit"
      value="submit"
    />
  </form>
</div>
`;

exports[`<Form /> Register Form renders a snapshot properly 1`] = `
<div>
  <h1
    className="title is-1"
  >
    Register
  </h1>
  <hr />
  <br />
  <form
    onSubmit={[Function]}
  >
    <div
      className="field"
    >
      <label
        className="label is-large"
      >
        Username
      </label>
      <div
        className="control has-icons-left"
      >
        <input
          className="input is-large"
          name="username"
          placeholder="Enter a username"
          required={true}
          type="text"
          value=""
        />
        <span
          className="icon is-small is-left"
        >
          <i
            className="fas fa-user"
          />
        </span>
      </div>
    </div>
    <div
      className="field"
    >
      <label
        className="label is-large"
      >
        Email
      </label>
      <div
        className="control has-icons-left"
      >
        <input
          className="input is-large"
          name="email"
          placeholder="Enter an email address"
          required={true}
          type="email"
          value=""
        />
        <span
          className="icon is-left"
        >
          <i
            className="fas fa-envelope"
          />
        </span>
      </div>
    </div>
    <div
      className="field"
    >
      <label
        className="label is-large"
      >
        Password
      </label>
      <div
        className="control has-icons-left"
      >
        <input
          className="input is-large"
          name="password"
          placeholder="Enter a password"
          required={true}
          type="password"
          value=""
        />
        <span
          className="icon is-left"
        >
          <i
            className="fas fa-lock"
          />
        </span>
      </div>
    </div>
    <br />
    <input
      className="button is-primary is-large is-fullwidth"
      type="submit"
      value="submit"
    />
  </form>
</div>
`;
